### PR TITLE
Post: Policy based design in C++

### PR DESCRIPTION
**Pull request by vmrob**

### CODE REVIEW

File: content/post/policy-based-design-cpp.md [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+draft = false
+date = "2015-09-01T19:05:13-06:00"
+tags = ["C++"]
+description = "Impementing the strategy pattern through the use of templatized mixins in C++"
+
++++
+
+### Design Patterns
+
+As software developers, we use design patterns to name and classify common programming idioms using a common set of jargon. It allows us to describe the higher level design of our code while providing a good idea of what a set of code accomplishes without having to understand everything about it.
```
Comment (ccbrown) [inline content/post/policy-based-design-cpp.md]: The title and first sentence or two often decide whether or not this thing actually gets read or not. Might want to consider a more to-the-point or thought provoking hook.

File: content/post/policy-based-design-cpp.md [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    // verify that calculate returns the policy's calculated result
+    assert(f.calculate() == mock_algorithm::expected);
+}
+
+void complex_algorithm_test() {
+    // testing for the complex algorithm
+}
+```
+
+Even if you have already been doing something similar in C++, hopefully you'll find this useful next time you need to stich together complex objects and still need to be able to test each component in isolation. Happy coding!
```
Comment (ccbrown) [inline content/post/policy-based-design-cpp.md]: "stich"

File: content/post/policy-based-design-cpp.md [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    }), str.end());
+
+std::cout << str << std::endl; // "FooBar"
+```
+
+In the above example, the logic of _what_ to erase has been separated from the logic of _how_ it is erased. The result is a completely reusable set of functions that can be mixed and matched to cover even more functionality.
+
+### Mixins
+
+Many OOP languages have the concept of [mixins](https://en.wikipedia.org/wiki/Mixin) which promote the simplification of object design by reducing a class to a series of reusable data and behavior. In php, these are called traits and can be used to keep classes small and tightly focused.
```
Comment (ccbrown) [inline content/post/policy-based-design-cpp.md]: capitalize "php"?

File: content/post/policy-based-design-cpp.md (hunk 1 above)
Comment (vmrob) [inline content/post/policy-based-design-cpp.md]: Good point.

File: content/post/policy-based-design-cpp.md (hunk 2 above)
Comment (vmrob) [inline content/post/policy-based-design-cpp.md]: Righto...

File: content/post/policy-based-design-cpp.md (hunk 3 above)
Comment (vmrob) [inline content/post/policy-based-design-cpp.md]: Yeah. I should. Same with other things we normally use in a lowercase context like Git.

**Comment by ccbrown** [comment]
I really hate GitHub for this.

**Comment by vmrob** [comment]
I hate github for this because of the inability to draft a bunch of comments and then post them all at once.

**Comment by vmrob** [comment]
I've reduced the article a good bit. I think it's a little more focused now; what do you think?

File: content/post/policy-based-design-cpp.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+int main() {
+    printer<stateless> p1;
+    printer<stateful> p2;
+
+    printf("%lu\n", sizeof(p1)); // 1
+    printf("%lu\n", sizeof(p2)); // 8
+}
+```
+
+What might be the strongest benefit to this design is that it greatly simplifies unit tests. The ability to swap in mock-policies and test subclasses in the absence of the parent class complexities is what really what this is all about.
```
Comment (ccbrown) [inline content/post/policy-based-design-cpp.md]: "is what really what"

**Pull request closed without merging**